### PR TITLE
perf(router): circuit break ingester connections

### PR DESCRIPTION
**Pull request by domodwyer**
Circuit-break ingester connections in the routers, so that we don't send requests to unavailable ingesters during deployments.

Removes the hack I added in https://github.com/influxdata/influxdb_iox/pull/6509, pulling the balancing logic "up" into the IOx codebase again. This should improve the latency & error rate during deployments.

Part of https://github.com/influxdata/influxdb_iox/issues/6173 (only metrics outstanding, coming next).

---

* perf(router): circuit break ingester connections (7596dc082)
      
      Adds on-path health checking / recording using the CircuitBreaker
      construct, stopping requests to unhealthy upstreams (minus the probe
      requests) until they recover.
      
      This removes the horrible gRPC balancer hack I added to get us deployed
      ASAP, and should eliminate latency spikes and elevated error responses
      observed during deployments as a result.

### CODE REVIEW

File: router/src/dml_handlers/rpc_write.rs [hunk 1]
```diff
@@ -1,73 +1,25 @@
...
-                    //
-                    // Do not leak this task, or panic trying to send on the
-                    // closed channel.
-                    return;
-                }
-            }
-
-            tokio::time::sleep(Duration::from_secs(5)).await;
-        }
-    });
```
Comment (domodwyer) [inline router/src/dml_handlers/rpc_write.rs:66]: :wave:

File: router/src/dml_handlers/rpc_write/balancer.rs [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    /// an infinite / cycling iterator. A node that becomes unavailable after
+    /// the snapshot was taken will continue to be returned by the iterator.
+    pub(super) fn endpoints(&self) -> impl Iterator<Item = &'_ CircuitBreakingClient<T, C>> {
+        // Grab and increment the current counter.
+        let counter = COUNTER.with(|cell| {
+            let mut cell = cell.borrow_mut();
+            let new_value = cell.wrapping_add(1);
+            *cell = new_value;
+            new_value
+        });
```
Comment (domodwyer) [inline router/src/dml_handlers/rpc_write/balancer.rs:58]: This is the same logic in the `RoundRobin` sharder that we used to use for this:

https://github.com/influxdata/influxdb_iox/blob/7596dc0826d49b162206d3ab104952ca5b68d085/sharder/src/round_robin.rs#L33-L48

File: router/src/dml_handlers/rpc_write/circuit_breaking_client.rs [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+/// A thin composite type decorating the [`WriteClient`] functionality of `T`,
+/// with circuit breaking logic from [`CircuitBreaker`].
+#[derive(Debug)]
+pub(super) struct CircuitBreakingClient<T, C = CircuitBreaker> {
+    /// The underlying [`WriteClient`] implementation.
+    inner: T,
+    /// The circuit-breaking logic.
+    state: C,
+}
```
Comment (domodwyer) [inline router/src/dml_handlers/rpc_write/circuit_breaking_client.rs:36]: This composes together an RPC client and a `CircuitBreaker`.

File: router/src/dml_handlers/rpc_write/circuit_breaking_client.rs [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+impl CircuitBreakerState for CircuitBreaker {
+    fn is_usable(&self) -> bool {
+        self.is_healthy() || self.should_probe()
+    }
+
+    fn observe<T, E>(&self, r: &Result<T, E>) {
+        self.observe(r)
+    }
+}
```
Comment (domodwyer) [inline router/src/dml_handlers/rpc_write/circuit_breaking_client.rs:26]: Here we invoke the methods on the `CircuitBreaker` to determine if the request should be sent (@carols10cents)

File: router/src/dml_handlers/rpc_write/balancer.rs [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+        //
+        // This ensures unhealthy nodes are not continuously (and unnecessarily)
+        // polled/probed in the iter cycle below. The low frequency and impact
+        // of a node becoming unavailable during a single request easily
+        // outweighs the trade-off of the constant health evaluation overhead.
+        let snapshot = self
+            .endpoints
+            .iter()
+            .filter(|e| e.is_usable())
+            .collect::<Vec<_>>();
```
Comment (domodwyer) [inline router/src/dml_handlers/rpc_write/balancer.rs:70]: And here we build a list of "healthy" nodes, using `is_usable()` to filter out the unhealthy nodes.

File: router/src/dml_handlers/rpc_write.rs [hunk 6]
```diff
@@ -204,6 +150,29 @@ where
...
+            .next()
+            .ok_or(RpcWriteError::NoUpstream)?
+            .write(req.clone())
+            .await
+        {
+            Ok(()) => return Ok(()),
+            Err(e) => warn!(error=%e, "failed ingester rpc write"),
+        };
+        tokio::time::sleep(delay).await;
+        delay = delay.saturating_mul(2);
```
Comment (alamb) [inline router/src/dml_handlers/rpc_write.rs:172]: This looks like an exponential backoff loop -- typically I think such loops also have a max delay -- so like the delay doubles up to a point and then is capped -- maybe 5 or 10 seconds would be reasonable here

File: router/src/dml_handlers/rpc_write/balancer.rs [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+
+use super::{
+    circuit_breaker::CircuitBreaker,
+    circuit_breaking_client::{CircuitBreakerState, CircuitBreakingClient},
+};
+
+thread_local! {
+    /// A per-thread counter incremented once per call to
+    /// [`Balancer::endpoints()`].
+    static COUNTER: RefCell<usize> = RefCell::new(0);
```
Comment (alamb) [inline router/src/dml_handlers/rpc_write/balancer.rs:11]: Given the the load balancer is running as a tokio task (and thus in many different threads) I wonder if this will work as expected? 

It seems like an individual load balancer might see see the same value in the counter more than once, depending on which thread it was scheduled on. Since `#[tokio::test]` is run on a single thread such problems won't show up in tests.

Is there some reason this isn't a field directly on the `Balancer` itself?

File: router/src/dml_handlers/rpc_write/balancer.rs [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+            balancer
+                .endpoints()
+                .next()
+                .expect("should yield healthy client")
+                .write(WriteRequest::default())
+                .await
+                .expect("should succeed");
+        }
+
+        assert_eq!(circuit_err.ok_count(), 0);
```
Comment (alamb) [inline router/src/dml_handlers/rpc_write/balancer.rs:221]: Nice tests 👍 

One scenario I didn't see covered was how the load balancer handles the transition from unusable to usable

So for example,   at this point in the test perhaps set `circut_err.set_usable(true)` and then verify that the balancer subsequently sent traffic to it

File: router/src/dml_handlers/rpc_write.rs [hunk 9]
```diff
@@ -1,73 +1,25 @@
...
-                    // The gRPC balancer task has stopped - likely because
-                    // the server has stopped.
-                    //
-                    // Do not leak this task, or panic trying to send on the
-                    // closed channel.
-                    return;
-                }
-            }
-
-            tokio::time::sleep(Duration::from_secs(5)).await;
```
Comment (alamb) [inline router/src/dml_handlers/rpc_write.rs:64]: Is the expected improved recovery performance largely due to the removal of this 5 second sleep?

**Review by alamb — APPROVED** [review]
I reviewed the code. I am not sure about a few things, but I am approving anyways to ensure we can merge this as quickly as possible

1. I think testing the recovery when the balancer recovery logic  
1. I don't understand how the thread local COUNTER works with a multithreaded tokio executor

File: ioxd_router/src/lib.rs [hunk 10]
```diff
@@ -259,14 +263,17 @@ pub async fn create_router2_server_type(
...
-    let ingester_addresses = router_config.ingester_addresses.join(",");
-
-    let grpc_connections = router::dml_handlers::build_ingester_connection(
-        ingester_addresses.split(',').map(|s| format!("http://{s}")),
-    );
+    let ingester_connections = router_config.ingester_addresses.join(",");
+    let ingester_connections = ingester_connections.split(',').map(|s| {
+        WriteServiceClient::new(
+            Channel::from_shared(format!("http://{s}"))
+                .expect("invalid ingester connection address")
```
[the remark this answers is not in the document]
Comment (domodwyer) [reply ioxd_router/src/lib.rs:270]: I'm sure we'll get there! This splitting/joining shouldn't exist at all - it should be done by the clap variable population itself, there's a ticket somewhere to fix it.

File: router/src/dml_handlers/rpc_write.rs (hunk 6 above)
Comment (domodwyer) [reply router/src/dml_handlers/rpc_write.rs:172]: It is already:

https://github.com/influxdata/influxdb_iox/blob/7596dc0826d49b162206d3ab104952ca5b68d085/router/src/dml_handlers/rpc_write.rs#L120

File: router/src/dml_handlers/rpc_write/balancer.rs (hunk 7 above)
Comment (domodwyer) [reply router/src/dml_handlers/rpc_write/balancer.rs:11]: > It seems like an individual load balancer might see see the same value in the counter more than once, depending on which thread it was scheduled on

Absolutely, but given that the threads don't process in lock-step and therefore they don't all use the same counter/upstream at the same time, I don't think this is a problem. The load is evenly distributed with a large enough number of requests - it is what we have been using for a while already, I just inlined it :+1:

File: router/src/dml_handlers/rpc_write/balancer.rs (hunk 8 above)
Comment (domodwyer) [reply router/src/dml_handlers/rpc_write/balancer.rs:221]: Good call, I shall add one in a minute :+1:

File: router/src/dml_handlers/rpc_write.rs (hunk 9 above)
Comment (domodwyer) [reply router/src/dml_handlers/rpc_write.rs:64]: The performance improvement stems from the fact the gRPC balancer does not consider an error RPC response from an ingester as a signal it is unhealthy. This is the right thing for it to do because it is not at the "application" layer, but further down, providing a "networking" primitive. It can't reasonably expect all errors to be bad enough to blacklist the ingester (think "not found" errors in a generic RPC system, for example).

During graceful shutdown an ingester returns the equivalent of `Err("i'm shutting down")`, but the gRPC balancer doesn't understand this, and keeps sending writes to it anyway. This PR fixes that by interpreting any failure to write data as an error that (with sufficient error rates) should cause us to temporarily blacklist the ingester.

TL;DR: we'll stop sending requests that we know won't work

Comment (domodwyer) [reply router/src/dml_handlers/rpc_write.rs:64]: (there's also other advantages, but this is the clearest one to reason about, and is definitely good enough to justify the change)
